File: app/src/main/java/top/pythong/contributioncalendar/GithubClient.kt
package top.pythong.contributioncalendar

import org.jsoup.Jsoup


/**
 * @author ChangJiahong
 * @date 2021/1/2
 * Create By Android Studio
 */
object GithubClient {
    fun parseFill(f: String): String = when (f) {
        "var(--color-calendar-graph-day-bg)" -> "def"
        "var(--color-calendar-graph-day-L1-bg)" -> "L1"
        "var(--color-calendar-graph-day-L2-bg)" -> "L2"
        "var(--color-calendar-graph-day-L3-bg)" -> "L3"
        "var(--color-calendar-graph-day-L4-bg)" -> "L4"
        else -> "null"
    }

    private fun parseDate(date: String, isEn: Boolean): String =
        when (date.split("-").toTypedArray()[1]) {
            "01" -> if (isEn) "Jan" else "一月"
            "02" -> if (isEn) "Feb" else "二月"
            "03" -> if (isEn) "Mar" else "三月"
            "04" -> if (isEn) "Apr" else "四月"
            "05" -> if (isEn) "May" else "五月"
            "06" -> if (isEn) "Jun" else "六月"
            "07" -> if (isEn) "Jul" else "七月"
            "08" -> if (isEn) "Aug" else "八月"
            "09" -> if (isEn) "Sep" else "九月"
            "10" -> if (isEn) "Oct" else "十月"
            "11" -> if (isEn) "Nov" else "十一"
            "12" -> if (isEn) "Dec" else "十二"
            else -> " "
        }

    fun getSheetByGithub(url:String): CalendarSheet {

        val doc = Jsoup.connect(url).get()
        val svgs = doc.getElementsByClass("js-calendar-graph-svg")[0].children()[0].children()
        val data: ArrayList<Week> = ArrayList()
        val monthTitles: ArrayList<CTitle> = ArrayList()
        var xIndex = 0 //列号
        var mTileFlag = false //记录月份标志
        for (gt in svgs) {
            when (gt.nodeName()) {
                "g" -> {
                    // 分割列数据
                    val col = gt;
                    val days = col.children()
                    val week = Week(xIndex++, days = ArrayList())
                    val p = parseDate(
                        days[0].attr("data-date"),
                        true
                    ) // 当前列第一天的月份

                    var f = false
                    for ((yIndex, day) in days.withIndex()) {
                        val wd = Day(
                            index = yIndex,
                            date = day.attr("data-date"),
                            dataCount = day.attr("data-count").toInt(),
                            fill = parseFill(day.attr("fill"))
                        )
                        val date = parseDate(wd.date, true)
                        if (date != p) {// 与当前列第一天不是同一月
                            f = true;
                        }
                        week.days.add(wd)
                    }
                    if (mTileFlag || week.days[0].date.split("-")[2] == "01") {
                        // 记录标题
                        monthTitles.add(
                            CTitle(
                                index = week.index, title = parseDate(week.days[0].date, true)
                            )
                        )
                        mTileFlag = false
                    }
                    mTileFlag = f
                    data.add(week)
                }
                "text" -> {
                    when (gt.className()) {
                        "month" -> {
                            // monthTitles.add(CTitle(index: ((int.parse(gt.attributes["x"])-1)/15 - 1).round(), title: gt.text));
                        }
                        "wday" -> {
                        }
                    }
                }
            }

        }

        return CalendarSheet(data = data, mTitle = monthTitles);
    }


    fun createData(): CalendarSheet {
        val weeks = arrayListOf(
            Week(
                0, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L2"),
                    Day(3, "", 0, "L3"),
                    Day(4, "", 0, "L4"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L2")
                )
            ),
            Week(
                1, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                2, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                3, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                4, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                5, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                6, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                7, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                8, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                9, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                10, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                11, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L1"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L1"),
                    Day(5, "", 0, "L1"),
                    Day(6, "", 0, "L1")
                )
            ),
            Week(
                12, arrayListOf(
                    Day(0, "", 0, "L1"),
                    Day(1, "", 0, "L1"),
                    Day(2, "", 0, "L3"),
                    Day(3, "", 0, "L1"),
                    Day(4, "", 0, "L2"),
//                Day(5, "", 0, "L1"),
//                Day(6, "", 0, "L1")
                )
            )
        )
        val monthTitles = arrayListOf(CTitle(0, "dec"), CTitle(4, "tep"))
        val data = CalendarSheet(weeks, monthTitles)
        return data
    }

}

File: calendar/lib/GithubClient.dart
import 'dart:convert';
import 'dart:io';
import 'package:calendar/Calendar.dart';
import 'package:calendar/HttpError.dart';
import 'package:dio/dio.dart';
import 'package:flutter/material.dart' as prefix;
import 'package:html/parser.dart' show parse;
import 'package:xpath_parse/xpath_selector.dart';

///
///获取GitHub贡献日历
///@author ChangJiahong
///@date 2020/12/30
///Create By Android Studio
///

class CalendarSheet{
  List<Week> data;
  List<CTitle> mTitle;
  CalendarSheet({this.data, this.mTitle});

}

class GithubClient {
  var httpClient = new HttpClient();

  static String parseFill(String f) {
    switch (f) {
      case "var(--color-calendar-graph-day-bg)":
        return "def";
      case "var(--color-calendar-graph-day-L1-bg)":
        return "L1";
      case "var(--color-calendar-graph-day-L2-bg)":
        return "L2";
      case "var(--color-calendar-graph-day-L3-bg)":
        return "L3";
      case "var(--color-calendar-graph-day-L4-bg)":
        return "L4";
    }
  }

  static String parseDate(String date, bool isEn){
    var mm = date.split("-")[1];
    switch(mm){
      case "01":
        return isEn?"Jan":"一月";
      case "02":
        return isEn?"Feb":"二月";
      case "03":
        return isEn?"Mar":"三月";
      case "04":
        return isEn?"Apr":"四月";
      case "05":
        return isEn?"May":"五月";
      case "06":
        return isEn?"Jun":"六月";
      case "07":
        return isEn?"Jul":"七月";
      case "08":
        return isEn?"Aug":"八月";
      case "09":
        return isEn?"Sep":"九月";
      case "10":
        return isEn?"Oct":"十月";
      case "11":
        return isEn?"Nov":"十一";
      case "12":
        return isEn?"Dec":"十二";
    }
  }


  static Future<CalendarSheet> getGithubCalendar(String url) async {
    Response res ;
    try {
      res = await new Dio().get(url);
    }catch(e){
      throw HttpError();
    }
    var html = res.data;
    // var list = XPath.source(html).query("//*[@id=\"js-pjax-container\"]/div[2]/div/div[2]/div[2]/div/div[2]/div[1]/div/div/div[1]").list();
    var document = parse(html);
    var svg = document.getElementsByClassName("js-calendar-graph-svg").first;

    // 取到所有的列数据
    var g_text = svg.children.first.children;
    List<Week> data = List();
    List<CTitle> monthTitles= List();
    int xIndex = 0;//列号
    bool mTileFlag=false;//记录月份标志
    var tempMonth="";
    for (var gt in g_text) {
      switch (gt.localName) {
        case "g":
          // 分割列数据
          var col = gt;
          var days = col.children;
          var week = Week(x:xIndex++);
          week.days=List();
          int yIndex = 0;//行号
          // print("-----------${week.x}--------");
          var p=parseDate(days[0].attributes["data-date"], true);// 当前列第一天的月份
          var f=false;
          for (var day in days) {
            var wd = Day(
                y: yIndex++,
                date: day.attributes["data-date"],
                dataCount: int.parse(day.attributes["data-count"]),
                fill: parseFill(day.attributes["fill"]));
            var date = parseDate(wd.date, true);
            if(date!=p){// 与当前列第一天不是同一月
              f=true;
            }

            week.days.add(wd);
            // print(day.outerHtml);
            // print(wd);
          }
          if(mTileFlag||week.days[0].date.split("-")[2]=="01") {
            // 记录标题
            monthTitles.add(CTitle(
                index: week.x, title: parseDate(week.days[0].date, true)));
            mTileFlag=false;
          }
          mTileFlag=f;
          data.add(week);
          break;
        case "text":
          switch(gt.className){
            case "month":
              // monthTitles.add(CTitle(index: ((int.parse(gt.attributes["x"])-1)/15 - 1).round(), title: gt.text));
              break;
            case "wday":

              break;
          }
          break;
      }

    }

    return CalendarSheet(data: data, mTitle: monthTitles);
  }
}
